### PR TITLE
CLDR-16390 Bailey explainer

### PR DESCRIPTION
**Pull request by srl295**
- update the implementations behind getPathsForValue to keep going when a hard value is hit. Won't affect getSourceLocale, etc, return values,
but will populate the tracing list with additional values.

- [ ] This PR completes the ticket

CLDR-16390

ALLOW_MANY_COMMITS=true

-----
Example output:

```
• //ldml/localeDisplayNames/territories/territory[@type="PM"]
• 576f1c82a03a22a8
• Value: Saint-Pierre-et-Miquelon
• Inheritance chain:
  • value: en_CA://ldml/localeDisplayNames/territories/territory[@type="PM"]
  • value: en://ldml/localeDisplayNames/territories/territory[@type="PM"]
  • none: root://ldml/localeDisplayNames/territories/territory[@type="PM"]
  • codeFallback: //ldml/localeDisplayNames/territories/territory[@type="PM"]
```

### CODE REVIEW

**Comment by srl295** [comment]
test needs update

File: tools/cldr-code/src/main/java/org/unicode/cldr/util/XMLSource.java [hunk 1]
```diff
@@ -590,6 +590,7 @@ public String getSourceLocaleIdExtended(
                 list.add(
                         new LocaleInheritanceInfo(
                                 locale, path, LocaleInheritanceInfo.Reason.value));
+                // Since we’re not resolving, there’s no way to look for a Bailey value here.
```
Comment (macchiati) [inline tools/cldr-code/src/main/java/org/unicode/cldr/util/XMLSource.java:593]: We should throw an exception if we try for a bailey value with a non-resolving source.

Comment (srl295) [reply tools/cldr-code/src/main/java/org/unicode/cldr/util/XMLSource.java:593]: Good point. and i'll add an exception up in `getPathsWhereFound()`, certainly.  This code, however is two levels deeper, in the XMLSource override.

Comment (macchiati) [reply tools/cldr-code/src/main/java/org/unicode/cldr/util/XMLSource.java:593]: ok. Are you going to update the test as part of this PR?

Comment (srl295) [reply tools/cldr-code/src/main/java/org/unicode/cldr/util/XMLSource.java:593]: Yes, I'll probably work on that Monday.

Comment (srl295) [reply tools/cldr-code/src/main/java/org/unicode/cldr/util/XMLSource.java:593]: Test found some bugs !

**Comment by srl295** [comment]
should pass tests now

File: tools/cldr-code/src/main/java/org/unicode/cldr/util/XMLSource.java (hunk 1 above)
Comment (srl295) [reply tools/cldr-code/src/main/java/org/unicode/cldr/util/XMLSource.java:593]: throws exception, passes tests

**Comment by macchiati** [comment]
I don't think it is quite working yet. I loaded it up on cldr-staging, and looked at https://cldr-staging.unicode.org/cldr-apps/v#/de/CompoundUnits/3f5664e23b5fc783
What I see is the following:
Screenshot 2023-05-01 at 13 37 31

**Comment by srl295** [comment]
> I don't think it is quite working yet. I loaded it up on cldr-staging, and looked at https://cldr-staging.unicode.org/cldr-apps/v#/de/CompoundUnits/3f5664e23b5fc783
> What I see is the following:
> Screenshot 2023-05-01 at 13 37 31

What should it do?

Some of the Jump buttons don't work because they go to non-existent paths.

**Comment by macchiati** [comment]
Some issues

1. The original path is not shown. That should be at the top (it was, in an earlier iteration). Without that, it is hard to follow.

- //ldml/units/unitLength[@type="long"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="one"][@gender="masculine"][@case="accusative"]

2. Some jumps don't work, because there is nothing to jump to. All of the following are really one transaction; picking up an alias in Root, changing the path, then accessing German. Only the **last** [Jump] (for German) has any meaning.

- Root [Jump]
- An alias was found at this location [Jump]
- //ldml/units/unitLength[@type="short"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="one"][@gender="masculine"][@case="accusative"]
- German [Jump]

3. The first two should be combined (aliases are only in Root)

- Root — An alias changes the path to:
- //ldml/units/unitLength[@type="short"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="one"][@gender="masculine"][@case="accusative"]

3a. It would be *much* easier to follow if we bolded the difference between the paths. Eg **[@count="one"]**. That can follow later, however.

4. We are not seeing the subsequent changes to the path (before code fallback) (I think) because part of the fallback is done in of the fallback is done in CLDRFile when it should be in XMLFile, namely by
    public String getCountPathWithFallback(String xpath, Count count, boolean winning) {
[mea culpa, by the way]

We should move that logic down into XMLSource eventually. But in the meantime you might be able to instrument that part also.

**Comment by srl295** [comment]
Generally, could we consider merging this when the lower level API is getting the right data, and look at UI improvements separately as a lot of low level code is touched here? (item #1 is now solved)

> Some issues
> 
> 1. The original path is not shown. That should be at the top (it was, in an earlier iteration). Without that, it is hard to follow.

I've restored that in a commit just now.

> * //ldml/units/unitLength[@type="long"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="one"][@gender="masculine"][@case="accusative"]
> 
> 2. Some jumps don't work, because there is nothing to jump to. All of the following are really one transaction; picking up an alias in Root, changing the path, then accessing German. Only the **last** [Jump] (for German) has any meaning.
> 
> * Root [Jump]
> * An alias was found at this location [Jump]
> * //ldml/units/unitLength[@type="short"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="one"][@gender="masculine"][@case="accusative"]
> * German [Jump]

Yes, that’s exactly why I opened #2879, so that the Jump buttons could be made reactive. 

> 3. The first two should be combined (aliases are only in Root)
> 
> * Root — An alias changes the path to:
> * //ldml/units/unitLength[@type="short"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="one"][@gender="masculine"][@case="accusative"]
> 
> 3a. It would be _much_ easier to follow if we bolded the difference between the paths. Eg **[@count="one"]**. That can follow later, however.

That's a good idea, show the edit difference between them.

Generally speaking, XPaths aren't what we actually show in most of the UI.  We could perhaps show PathHeader instead with an option to show the XPath. 

> 4. We are not seeing the subsequent changes to the path (before code fallback) (I think) because part of the fallback is done in of the fallback is done in CLDRFile when it should be in XMLFile, namely by
>    public String getCountPathWithFallback(String xpath, Count count, boolean winning) {
>    [mea culpa, by the way]
>
> We should move that logic down into XMLSource eventually. But in the meantime you might be able to instrument that part also.

I can instrument where the xpath changes as a result of getFallbackPaths, that yields the following: (note the `| fallback` entry marked with a `»`)

```
• //ldml/units/unitLength[@type="long"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="one"][@gender="feminine"][@case="locative"][@alt="variant"]
• 2c01af8246eb4f09
• Value: {0}²
• Inheritance chain:
Path: //ldml/units/unitLength[@type="short"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="one"], loc: cs
  • none: cs://ldml/units/unitLength[@type="long"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="one"][@gender="feminine"][@case="locative"][@alt="variant"]
  • none: root://ldml/units/unitLength[@type="long"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="one"][@gender="feminine"][@case="locative"][@alt="variant"]
  | alias: root://ldml/units/unitLength[@type="short"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="one"][@gender="feminine"][@case="locative"][@alt="variant"]
  • none: cs://ldml/units/unitLength[@type="short"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="one"][@gender="feminine"][@case="locative"][@alt="variant"]
  • none: root://ldml/units/unitLength[@type="short"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="one"][@gender="feminine"][@case="locative"][@alt="variant"]
  | removedAttribute: //ldml/units/unitLength[@type="short"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="one"][@gender="feminine"][@case="locative"]
    attribute=alt
  • none: cs://ldml/units/unitLength[@type="short"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="one"][@gender="feminine"][@case="locative"]
  • none: root://ldml/units/unitLength[@type="short"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="one"][@gender="feminine"][@case="locative"]
  | changedAttribute: //ldml/units/unitLength[@type="short"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="other"][@gender="feminine"][@case="locative"]
    attribute=count
  • none: cs://ldml/units/unitLength[@type="short"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="other"][@gender="feminine"][@case="locative"]
  • none: root://ldml/units/unitLength[@type="short"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="other"][@gender="feminine"][@case="locative"]
  • codeFallback: //ldml/units/unitLength[@type="short"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="other"][@gender="feminine"][@case="locative"]
» | fallback: cs://ldml/units/unitLength[@type="long"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="one"][@gender="feminine"][@case="locative"][@alt="variant"]
  • none: cs://ldml/units/unitLength[@type="long"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="one"][@alt="variant"]
  • none: root://ldml/units/unitLength[@type="long"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="one"][@alt="variant"]
  | alias: root://ldml/units/unitLength[@type="short"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="one"][@alt="variant"]
  • none: cs://ldml/units/unitLength[@type="short"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="one"][@alt="variant"]
  • none: root://ldml/units/unitLength[@type="short"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="one"][@alt="variant"]
  | removedAttribute: //ldml/units/unitLength[@type="short"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="one"]
    attribute=alt
  • value: cs://ldml/units/unitLength[@type="short"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="one"]
    /Users/srl295/src/cldr/common/main/cs.xml:16807:66:  XML Source
  • none: root://ldml/units/unitLength[@type="short"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="one"]
  | changedAttribute: //ldml/units/unitLength[@type="short"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="other"]
    attribute=count
  • value: cs://ldml/units/unitLength[@type="short"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="other"]
    /Users/srl295/src/cldr/common/main/cs.xml:16810:68:  XML Source
  • value: root://ldml/units/unitLength[@type="short"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="other"]
    /Users/srl295/src/cldr/common/main/root.xml:4830:68:  XML Source
  • codeFallback: //ldml/units/unitLength[@type="short"]/compoundUnit[@type="power2"]/compoundUnitPattern1[@count="other"]
```

**Comment by srl295** [comment]
> > Generally speaking, XPaths aren't what we actually show in most of the UI. We could perhaps show PathHeader instead with an option to show the XPath.
> 
> Yes, that should be the end goal for non-power users. We could have a hover with the path, that would be useful for us.
> 
> > I can instrument where the xpath changes as a result of getFallbackPaths, that yields the following: (note the | fallback entry marked with a...
> > Excellent! That shows us what is there — _**and**_ what we need to fix!

OK, I've committed this. I just call it `fallback` but it could probably have a better name.

**Comment by macchiati** [comment]
I put together a chart showing what is happening (there might be a missing step in your output), and what I think ought to happen (PROPOSED)
https://docs.google.com/spreadsheets/d/1ho6La6QlfqpSdC9kNBaWvKmZYRA_9QUXaDQXEggnZl4

**Comment by srl295** [comment]
> I put together a chart showing what is happening (there might be a missing step in your output), and what I think ought to happen (PROPOSED)
> https://docs.google.com/spreadsheets/d/1ho6La6QlfqpSdC9kNBaWvKmZYRA_9QUXaDQXEggnZl4

OK. it's not tracing through getFallbackPaths() just noting the final output.

**Comment by macchiati** [comment]
That's ok for it not to trace through getFallbackPaths: I think we know what is happening there. What do you think about the PROPOSED? (Will of course have to run it by others, and it shouldn't block this PR).

**Comment by srl295** [comment]
> That's ok for it not to trace through getFallbackPaths: I think we know what is happening there. What do you think about the PROPOSED? (Will of course have to run it by others, and it shouldn't block this PR).

It makes some sense, I would have to look at the data to see what would be impacted.   Dropping the alt at first is definitely a big change in some sense.

**Review by btangmu — CHANGES_REQUESTED** [review]

File: tools/cldr-code/src/main/java/org/unicode/cldr/util/XMLSource.java [hunk 2]
```diff
@@ -1089,6 +1090,11 @@ private AliasLocation getPathLocation(
                 boolean skipFirst,
                 boolean skipInheritanceMarker,
                 List<LocaleInheritanceInfo> list) {
+
+            //   When calculating the Bailey values, we track the final
+            //   return value as firstValue. If non-null, this will become
+            //   the function's ultimate return value.
```
[the remark this answers is not in the document]
Comment (srl295) [reply tools/cldr-code/src/main/java/org/unicode/cldr/util/XMLSource.java:1096]: I'm not sure, looks like spotless did it

File: tools/cldr-code/src/main/java/org/unicode/cldr/util/CLDRFile.java [hunk 3]
```diff
@@ -554,32 +554,47 @@ public String getBaileyValue(
...
+                    "getPathsWhereFound() is only valid on a resolved CLDRFile");
+        }
+        LinkedList<LocaleInheritanceInfo> list = new LinkedList<>();
         // first, call getSourceLocaleIdExtended to populate the list
         Status status = new Status();
-        getSourceLocaleIdExtended(xpath, status, true, list);
+        final String locale1 = // for debugging
+                getSourceLocaleIdExtended(xpath, status, false, list);
         final String path1 = status.pathWhereFound;
+        System.out.println("Path: " + path1 + ", loc: " + locale1);
```
[the remark this answers is not in the document]
Comment (srl295) [reply tools/cldr-code/src/main/java/org/unicode/cldr/util/CLDRFile.java]: I think I've removed that now, thanks

File: tools/cldr-code/src/main/java/org/unicode/cldr/util/XMLSource.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (srl295) [reply tools/cldr-code/src/main/java/org/unicode/cldr/util/XMLSource.java:1096]: when in Rome (or when in XMLSource) … see the lines just prior:
image

File: tools/cldr-code/src/main/java/org/unicode/cldr/util/CLDRFile.java [hunk 4]
```diff
@@ -554,32 +554,47 @@ public String getBaileyValue(
...
+     * is used to explain inheritance and bailey values. The list must be interpreted in order. When
+     * {@link LocaleInheritanceInfo.Reason#isTerminal()} return true, that indicates a successful
+     * lookup and partitions values from subsequent bailey values.
      *
-     * @return map from xpath to locale of one or more contributing paths in the case of glossonymns
-     *     or other constructed values.
+     * @see #getBaileyValue(String, Output, Output)
+     * @see #getSourceLocaleIdExtended(String, Status, boolean)
      */
     public List<LocaleInheritanceInfo> getPathsWhereFound(String xpath) {
```
[the remark this answers is not in the document]
Comment (srl295) [reply tools/cldr-code/src/main/java/org/unicode/cldr/util/CLDRFile.java:565]: Correct, however, `ResolvingSource.getBaileyValue()` actually just calls `ResolvingSource.getPathLocation(…,  /*skipFirst*/ true)` to skip the first value.  That same `getPathLocation()` is the one that does the heavy lifting to build up the list.

[the remark this answers is not in the document]
Comment (srl295) [reply tools/cldr-code/src/main/java/org/unicode/cldr/util/CLDRFile.java:565]: getFallbackPath() should be called from XMLSource - that's mentioned in https://unicode-org.atlassian.net/browse/CLDR-13266 - when that is so and the glossonym stuff cleaned up, CLDRFile.getBaileyValue() can just call XMLSource.getBaileyValue() which will be the same code path.

We could also add tests to show comparable results. `getPathsWhereFound()` does a lot more work that `getBaileyValue()` might not want or need.

**Comment by macchiati** [comment]
SGTM

**Review by macchiati — APPROVED** [review]

**Comment by srl295** [comment]
@btangmu can we approve and do testing in a follow on?

**Comment by srl295** [comment]
Api doesn't lend itself to cross testing versus Bailey especially for constructed.

**Review by btangmu — APPROVED** [review]
a follow-up is needed to ensure getPathsWhereFound and getBaileyValue actually implement the same algorithm and always find the same localeWhereFound and pathWhereFound; one way would be for getBaileyValue to be a short wrapper around getPathsWhereFound

**Comment by srl295** [comment]
> "can we approve and do testing in a follow on?" -- yes, though more than just testing is needed -- it's problematic to have two separate functions getPathsWhereFound and getBaileyValue that supposedly implement the same crucial algorithm. If there's a performance problem with the extra explanatory work done by getPathsWhereFound, that extra work could be contingent on a boolean or something. Sooner or later -- probably sooner, since there are progress-blocking inheritance bugs -- the algorithm is going to change, and it's going to be a burden to keep the two functions consistent

Actually, getPathsWhereFound does not add any new lookup code.  It adds tracing to the existing implementation of getSourceLocaleIdExtended, the same implementation that is called by getBaileyValue. 

> btw why no "This PR completes the ticket" checkbox for this PR? I'm approving the PR with the understanding that if there were a checkbox it wouldn't be checked

It must have been accidentally removed. I've re-added it, unchecked. Thanks for the catch.

**Comment by btangmu** [comment]
"Actually, getPathsWhereFound does not add any new lookup code. It adds tracing to the existing implementation of getSourceLocaleIdExtended, the same implementation that is called by getBaileyValue."

OK, still, what needs to be explained is getBaileyValue, which is 37 lines long and doesn't actually call getSourceLocaleIdExtended directly.

What we need is a "getBaileyValue explainer", not a "getSourceLocaleIdExtended explainer" :-)
